### ISSUE TITLE
Is there a way to mock Minio Client in Golang Unit tests?

### ISSUE THREAD

**Issue by ziranl16** [minio/minio-go#1497]
Hi, I am writing unit tests in Golang. I wonder how can I mock the Minio Client in Golang, any suggestions? Looking forward to your help.

**Comment by klauspost** [comment]
If you need a higher level, create an interface for the functions you need in the client and replace your client use with the interfaces and implement the mocking functionality you need.

If you need lower level, you can override the [Transport](https://github.com/minio/minio-go/blob/7f05f7e9774ec1860922ce5bf2f6717faf8677aa/api.go#L99) in the client (low level, complex).

**Comment by ziranl16** [comment]
Thanks a lot @klauspost ! I will try it out immediately.

**Comment by ziranl16** [comment]
Hi @klauspost , may I please double check, if I want to mock the makeBucket method. shall I define like this in my interface:
```MakeBucket(ctx context.Context, bucketName string, opts minio.MakeBucketOptions) (err error)``` ?
I need to use ```minio.MakeBucketOptions``` in my interface parameters, and will this kindof violates the OOP principles. I tried using my self-defined MakeBucketOptions which I copied from Minio-go, but it says that minio Client does not implement this method. Looking forward to ur help! : )

**Comment by klauspost** [comment]
@ziranl16 Great that Go is not an OOP language. To implement an interface you need to implement it with the same types. I will not go into Dependecy Injection and similar Go related stuff, you can google that.

TLDR; Create an interface, inject a `*minio.Client` or your mock as you see fit:

```
// Add functions you need to mock.
type ObjectStore interface() {
	MakeBucket(ctx context.Context, bucketName string, opts minio.MakeBucketOptions) error
	PutObject(ctx context.Context, bucketName, objectName string, reader io.Reader, objectSize int64,
	opts minio.PutObjectOptions) (info minio.UploadInfo, err error)
}

// Use to mock:

type moctObjectStore struct {
}

func (m moctObjectStore) MakeBucket(ctx context.Context, bucketName string, opts minio.MakeBucketOptions)  error {
	return nil
}

func (m moctObjectStore) PutObject(ctx context.Context, bucketName, objectName string, reader io.Reader, objectSize int64,
	opts minio.PutObjectOptions) (info minio.UploadInfo, err error) {
	return minio.UploadInfo{/* TODO: Fill */}, nil
}
```

**Comment by ziranl16** [comment]
Thanks a lot for your detailed explanation! Thanks a lot for your help, @klauspost !
